### PR TITLE
 Refactor autoscaler functionality into an interface. 

### LINKED ISSUE #2131: Allow disabling creation of the HPA in raw deployment mode
/kind feature

**Describe the solution you'd like**
Allow disabling creation of the HPA in raw deployment mode.

I would like to be able to control the scaling of the deployment created by my KServe inference service with an external tool ([KEDA](https://keda.sh/)). In order to support this use case I need a way to disable KServe's creation of the HPA. I was thinking adding a new autoscaler class called `none` which disables the creation of any resources by the autoscaler reconciler.

### PR DESCRIPTION
**Pull request by Jason4849**
**What this PR does / why we need it**:

Thanks @cmaddalozzo for your commit. I added test cases with your commit(#2132 ).
This PR makes it possible to disable the creation of an HPA on raw deployments by adding an annotation:
~~~
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  name: "my-model"
  annotations:
    "serving.kserve.io/autoscalerClass": "none"
~~~
Fixes #2131 

**Type of changes**
Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevent result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

**Test manager image**: quay.io/jooholee/manager-12782ae64d3f5f3dbe4595b05a2cb78d@sha256:f096b4b550d56a9efe064792d722efb6af82d32aad0b4c8a27797d88e29dc0c4

**Test Steps:**
~~~
git clone https://github.com/kserve/kserve.git
cd kserve

kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.11.0/cert-manager.yaml

cat <<EOF >./deploy-target-kserve.sh
#!/bin/bash
cd config/default; kustomize edit add resource certmanager/certificate.yaml; cd ../..
mode='{"defaultDeploymentMode":"RawDeployment"}' yq -i '.data.deploy=strenv(mode)' config/configmap/inferenceservice.yaml
kustomize build config/overlays/development | kubectl apply -f -
kubectl wait --for=condition=ready pod -l control-plane=kserve-controller-manager -n kserve --timeout=300s
kustomize build config/runtimes | kubectl apply -f -
EOF

cat <<EOF >./config/overlays/development/manager_image_patch.yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: kserve-controller-manager
  namespace: kserve
spec:
  template:
    spec:
      containers:
        - name: manager
          command:
          image: quay.io/jooholee/manager-12782ae64d3f5f3dbe4595b05a2cb78d@sha256:bed9514d7c8ede23ff52d5d3c13546377861b029d0d5025a770edb2896c393c3
EOF

chmod 777 ./deploy-target-kserve.sh
./deploy-target-kserve.sh

kubectl create ns kserve-test
kubectl config set-context --current --namespace=kserve-test

cat <<EOF| kubectl apply -f -
apiVersion: "serving.kserve.io/v1beta1"
kind: "InferenceService"
metadata:
  annotations:
    "serving.kserve.io/autoscalerClass": "none"
    "serving.kserve.io/deploymentMode": "RawDeployment"
  name: "sklearn-irisv2"
spec:
  predictor:
    model:
      modelFormat:
        name: sklearn
      runtime: kserve-mlserver
      storageUri: "gs://seldon-models/sklearn/mms/lr_model"
EOF
~~~
Check if HPA is not created

### CODE REVIEW

**Comment by Jason4849** [comment]
@pugangxa @alexagriffith could you please review this ?

**Comment by Jason4849** [comment]
I don't understand why the e2e test failed regarding tensorflow. anybody can take a look at this error?

**Comment by Jason4849** [comment]
Now this PR passed the tests and need reviewers.
could you please have a look at this? @pugangxa @alexagriffith @yuzisun @cliveseldon

**Comment by Jooho** [comment]
Any update on this PR?

**Comment by yuzisun** [comment]
> Any update on this PR?

@Jason4849 Sorry for the delay, will review this asap

File: pkg/controller/v1beta1/inferenceservice/reconcilers/autoscaler/autoscaler_reconciler_test.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	// "github.com/onsi/gomega"
+	// gomegaTypes "github.com/onsi/gomega/types"
+	// istiov1alpha3 "istio.io/api/networking/v1alpha3"
+	// "istio.io/client-go/pkg/apis/networking/v1alpha3"
+	// corev1 "k8s.io/api/core/v1"
+	// metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	// "knative.dev/pkg/apis"
+	// duckv1 "knative.dev/pkg/apis/duck/v1"
+	// "knative.dev/pkg/network"
+	// "net/url"
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/autoscaler/autoscaler_reconciler_test.go]: Do we need these dependencies here?

File: pkg/controller/v1beta1/inferenceservice/reconcilers/autoscaler/autoscaler_reconciler.go [hunk 2]
```diff
@@ -70,27 +81,21 @@ func getAutoscalerClass(metadata metav1.ObjectMeta) constants.AutoscalerClassTyp
 
 func createAutoscaler(client client.Client,
 	scheme *runtime.Scheme, componentMeta metav1.ObjectMeta,
-	componentExt *v1beta1.ComponentExtensionSpec) (*Autoscaler, error) {
-	as := &Autoscaler{}
+	componentExt *v1beta1.ComponentExtensionSpec) (Autoscaler, error) {
 	ac := getAutoscalerClass(componentMeta)
```
Comment (yuzisun) [inline pkg/controller/v1beta1/inferenceservice/reconcilers/autoscaler/autoscaler_reconciler.go:86]: We probably need a configuration for the default autoscaler type in addition to the annotation overrides

**Comment by lizzzcai** [comment]
is there any update on this PR? Will it be in `v0.11`? Thanks.

**Comment by yuzisun** [comment]
closing this in favor of https://github.com/kserve/kserve/pull/3086

**Comment by yuzisun** [comment]
> is there any update on this PR? Will it be in `v0.11`? Thanks.

We plan to include in 0.11.1

**Pull request closed without merging**